src/openstack_cloud/openstack_cloud.py

```diff
@@ -0,0 +1,606 @@
+
+
+class OpenstackCloud:
+    """Client to interact with OpenStack cloud.
+
+    The OpenStack server name is managed by this cloud. Caller refers to the instances via
+    instance_id. If the caller needs the server name, e.g., for logging, it can be queried with
+    get_server_name.
+    """
+
+    def __init__(self, clouds_config: dict[str, dict], cloud: str, prefix: str):
+        """Create the object.
+
+        Args:
+            clouds_config: The openstack clouds.yaml in dict format.
+            cloud: The name of cloud to use in the clouds.yaml.
+            prefix: Prefix attached to names of resource managed by this instance. Used for
+                identifying which resource belongs to this instance.
+        """
+        self._clouds_config = clouds_config
+        self._cloud = cloud
+        self.prefix = prefix
+
+    # Ignore "Too many arguments" as 6 args should be fine. Move to a dataclass is new args are
+    # added.
+    def launch_instance(  # pylint: disable=R0913
+        self, instance_id: str, image: str, flavor: str, network: str, userdata: str
+    ) -> OpenstackInstance:
+        """Create an OpenStack instance.
+
+        Args:
+            instance_id: The instance ID to form the instance name.
+            image: The image used to create the instance.
+            flavor: The flavor used to create the instance.
+            network: The network used to create the instance.
+            userdata: The cloud init userdata to startup the instance.
+
+        Raises:
+            OpenstackError: Unable to create OpenStack server.
+
+        Returns:
+            The OpenStack instance created.
+        """
+        full_name = self.get_server_name(instance_id)
+        logger.info("Creating openstack server with %s", full_name)
+
+        with _get_openstack_connection(
+            clouds_config=self._clouds_config, cloud=self._cloud
+        ) as conn:
+            security_group = OpenstackCloud._ensure_security_group(conn)
+            keypair = OpenstackCloud._setup_keypair(conn, full_name)
+
+            try:
+                server = conn.create_server(
+                    name=full_name,
+                    image=image,
+                    key_name=keypair.name,
+                    flavor=flavor,
+                    network=network,
+                    security_groups=[security_group.id],
+                    userdata=userdata,
+                    auto_ip=False,
+                    timeout=_CREATE_SERVER_TIMEOUT,
+                    wait=True,
+                )
+            except openstack.exceptions.ResourceTimeout as err:
+                logger.exception("Timeout creating openstack server %s", full_name)
+                logger.info(
+                    "Attempting clean up of openstack server %s that timeout during creation",
+                    full_name,
+                )
+                try:
+                    conn.delete_server(name_or_id=full_name, wait=True)
```

Comment:
I'm wondering if this should use `self.delete_instance` and let the exception be handled there as well, just like how on L200 there is `self._delete_keypair`. To reduce the load on openstack we could reuse the connection instance.

Reply:
Good idea. I have updated the code.